library/objective-c/EnvoyEngineImpl.m

```diff
@@ -10,6 +10,72 @@ static void ios_on_exit() {
   NSLog(@"[Envoy] library is exiting");
 }
 
+typedef struct {
+  __unsafe_unretained EnvoyHTTPFilter *filter;
+} ios_http_filter_context;
+
+// TODO: factor out conversion
+// TODO: move impl
+static envoy_data toManagedNativeString(NSString *s) {
+  size_t length = s.length;
+  uint8_t *native_string = (uint8_t *)safe_malloc(sizeof(uint8_t) * length);
+  memcpy(native_string, s.UTF8String, length);
+  envoy_data ret = {length, native_string, free, native_string};
+  return ret;
+}
+
+static EnvoyMutableHeaders *to_ios_mutable_headers(envoy_headers headers) {
+  NSMutableDictionary *headerDict = [NSMutableDictionary new];
+  for (envoy_header_size_t i = 0; i < headers.length; i++) {
+    envoy_header header = headers.headers[i];
+    NSString *headerKey = [[NSString alloc] initWithBytes:header.key.bytes
+                                                   length:header.key.length
+                                                 encoding:NSUTF8StringEncoding];
+    NSString *headerValue = [[NSString alloc] initWithBytes:header.value.bytes
+                                                     length:header.value.length
+                                                   encoding:NSUTF8StringEncoding];
+    NSMutableArray *headerValueList = headerDict[headerKey];
+    if (headerValueList == nil) {
+      headerValueList = [NSMutableArray new];
+      headerDict[headerKey] = headerValueList;
+    }
+    [headerValueList addObject:headerValue];
+  }
+  // The C envoy_headers struct can be released now because the headers have been copied.
+  release_envoy_headers(headers);
+  return headerDict;
+}
```

Comment:
Would be great to add tests for these bridging functions before merging

Reply:
We don't yet have a framework in place for testing at the bridge layer, though @junr03 was talking about exploring this.